### PR TITLE
[Pipelining] Make forward_outputs and loss_fn strictly typed

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138125
* __->__ #138779
* #131762
* #138134

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/stage.py [hunk 1]
```diff
@@ -554,7 +554,7 @@ def forward_one_chunk(
         fwd_chunk_id: int,
         args: Tuple[Any, ...],
         kwargs: Optional[Dict[str, Any]] = None,
-    ):
+    ) -> Tuple[Any]:
```
Comment (kwen2501) [inline torch/distributed/pipelining/stage.py:557]: Is this change for the typing of loss_fn?

**Review by kwen2501 — APPROVED** [review]
LGTM.

File: torch/distributed/pipelining/stage.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/stage.py:557]: my original motivation was to type this function so it is easier for internal PP code to work with the forward outputs.

specifically in the next PR I need to 'zip' the forward outputs with the recv_infos so i can assign info.buffer = prev_stage_output.detach() to bypass the send/recv for that stage for V schedule special case.

I then thought that this change is helpful overall and we were bad for not having typed our loss function before.

Now I have second thoughts.  The user can provide any output type for their model and they also provide their loss function, so they are responsible for making the loss function input match the model output, but we are not necessarily justified in requiring the model to output a tuple.  If the model wants to output a tensor, i think we have to let it.

So now I think i need to relax this typing and just deal with the ambiguity of forward_outputs in my other PR.

Thoughts?

**Comment by wconstab** [comment]
closing for now as I no longer think this is a good idea.
